### PR TITLE
22655 - Wire in reverse_invoice for consolidated invoices

### PR DESCRIPTION
**Pull request by seeker25**
*Issue #:*
https://github.com/bcgov/entity/issues/22655

Flows:
1. EFT Client is overdue, they never navigate to the consolidation payments screen (NSF lock flow and don't hit the next button on credit card) - no cfs_invoice is created for consolidated invoice and no invoice_reference is created. They pay by regular EFT and it works.
2. EFT Client is overdue, pay by credit card, creates and pays consolidated invoices and reverses the "regular" invoice, this should unlock the users.
3. EFT Client is overdue they start the pay by credit card which creates a consolidated invoice, they eventually pay by EFT instead.
The job is smart enough to realize they had a consolidated invoice, reverses that and completes payment on the "regular" invoice

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-pay license (Apache 2.0).

### CODE REVIEW

File: pay-api/src/pay_api/services/payment.py [hunk 1]
```diff
@@ -625,24 +625,21 @@ def create_consolidated_invoices_payment(consolidated_invoices: List[InvoiceMode
...
-                transaction_number=invoice_number,
+                transaction_number=invoice_number_no_prefix,
                 line_items=consolidated_line_items,
                 cfs_account=cfs_account)
 
-        invoice_number: str = invoice_response.get('invoice_number')
-
         for invoice in consolidated_invoices:
-            inv_ref: InvoiceReferenceModel = InvoiceReferenceModel.find_by_invoice_id_and_status(
+            inv_ref = InvoiceReferenceModel.find_by_invoice_id_and_status(
```
Comment (seeker25) [inline pay-api/src/pay_api/services/payment.py:633]: Reuse the invoice reference if possible

File: jobs/payment-jobs/tasks/cfs_create_invoice_task.py [hunk 2]
```diff
@@ -105,13 +105,8 @@ def _cancel_rs_invoices(cls):
...
                                                    invoice_reference.invoice_number)
-
-                    # Adjust to zero: -invoice.total + invoice.total = 0
-                    adjustment_negative_amount = -invoice.total
-                    CFSService.adjust_invoice(cfs_account=cfs_account,
-                                              inv_number=invoice_reference.invoice_number,
-                                              amount=adjustment_negative_amount)
-
+                    # This used to be adjust invoice, but the suggested way from Tara is to use reverse invoice.
+                    CFSService.reverse_invoice(invoice_reference.invoice_number)
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/cfs_create_invoice_task.py:109]: I tried this by hand on TEST, seems to work.. it does return

```
{
    "credit_memo": "11133",
    "CAS-Returned-Messages": "SUCCESS"
}
```

We don't typically do anything with the credit memo return though.

File: pay-api/src/pay_api/models/statement_settings.py [hunk 3]
```diff
@@ -14,14 +14,14 @@
...
-from sqlalchemy import ForeignKey
+from sqlalchemy import ForeignKey, exists
 
 from pay_api.utils.enums import StatementFrequency
 
 from .base_model import BaseModel
 from .db import db, ma
 from .payment_account import PaymentAccount
-
+from .statement import Statement
```
Comment (seeker25) [inline pay-api/src/pay_api/models/statement_settings.py:24]: I really don't like this, will move to a service

File: pay-api/src/pay_api/services/non_sufficient_funds.py [hunk 4]
```diff
@@ -120,25 +122,32 @@ def query_all_non_sufficient_funds_invoices(account_id: str):
...
-        invoice_list = converter.unstructure(invoice_search_model)
-        new_invoices = [converter.remove_nones(invoice_dict) for invoice_dict in invoice_list]
-
+        invoice_list = Converter().unstructure(invoice_search_model)
+        new_invoices = [Converter().remove_nones(invoice_dict) for invoice_dict in invoice_list]
+        statements = Converter().unstructure(StatementDTO.dao_to_dto(statements))
         data = {
             'total': total,
             'invoices': new_invoices,
+            'statements': statements,
```
Comment (seeker25) [inline pay-api/src/pay_api/services/non_sufficient_funds.py:150]: Need statements to show up on the NSF lock screen not invoices for EFT.

File: pay-api/src/pay_api/services/payment_account.py [hunk 5]
```diff
@@ -619,40 +619,50 @@ def create_account_event_payload(self, event_type: str, receipt_info: dict = Non
     @staticmethod
     def unlock_frozen_accounts(payment_id: int, payment_account_id: int):
         """Unlock frozen accounts."""
-        pay_account: PaymentAccount = PaymentAccount.find_by_id(payment_account_id)
+        pay_account = PaymentAccountModel.find_by_id(payment_account_id)
+        unlocked = False
         if pay_account.cfs_account_status == CfsAccountStatus.FREEZE.value:
```
Comment (seeker25) [inline pay-api/src/pay_api/services/payment_account.py:624]: Will use `pay_account.has_nsf_invoices`

File: jobs/payment-jobs/tests/jobs/test_generate_statements.py [hunk 6]
```diff
@@ -439,15 +463,19 @@ def test_gap_statements(session, test_name, admin_users_mock):
         assert statement.payment_methods != 'PAD,EFT'
         assert statement.payment_methods != 'EFT,PAD'
 
+    if test_name == 'interim_overlap':
```
Comment (seeker25) [inline jobs/payment-jobs/tests/jobs/test_generate_statements.py:466]: need a bit better assertions here

File: pay-api/src/pay_api/models/statement.py [hunk 7]
```diff
@@ -13,9 +13,11 @@
 # limitations under the License.
```
Comment (seeker25) [inline pay-api/src/pay_api/models/statement.py:13]: Merge went wrong

File: pay-api/src/pay_api/services/payment.py [hunk 8]
```diff
@@ -656,32 +663,45 @@ def create_consolidated_invoices_payment(consolidated_invoices: List[InvoiceMode
...
+                    CFSService.reverse_invoice(inv_number=invoice_reference.invoice_number)
+                    break
+            # Don't reverse original invoice here, we need to do so after receiving payment, otherwised we'll have a
+            # consolidated invoice reference active while the regular invoice is reversed. (Scenario where they don't
+            # go through the CC NSF process) This doesn't work well for our EFT job.
             consolidated_invoices.append(invoice)
             invoice_total += invoice.total - invoice.paid
             consolidated_line_items.append(*invoice.payment_line_items)
 
         payment, _ = cls.create_consolidated_invoices_payment(consolidated_invoices, consolidated_line_items,
```
Comment (seeker25) [inline pay-api/src/pay_api/services/payment.py:697]: Will refactor this in a bit, way too many params

File: pay-api/src/pay_api/services/payment_account.py [hunk 9]
```diff
@@ -606,42 +607,61 @@ def create_account_event_payload(self, event_type: str, receipt_info: dict = Non
...
+        elif pay_account.has_overdue_invoices:
+            # Reverse original invoices here, because users can still cancel out of CC payment process and pay via EFT.
+            # Note we do the opposite of this in the EFT task.
+            invoice_references = InvoiceReferenceModel.query \
+                .filter(InvoiceReferenceModel.invoice_number == invoice_number) \
+                .distinct(InvoiceReferenceModel.invoice_number) \
+                .all()
+            for invoice_reference in invoice_references:
+                if '-C' not in invoice_reference.invoice_number:
+                    #  TODO which invoice_reference status are we filtering here?
```
Comment (seeker25) [inline pay-api/src/pay_api/services/payment_account.py:633]: todo look at this

**Comment by seeker25** [comment]
unit tests + CI fixes to go

**Comment by seeker25** [comment]
Flows:

1. EFT Client is overdue, they never navigate to the consolidation payments screen (NSF lock flow and don't hit the next button on credit card) - no cfs_invoice is created for consolidated invoice and no invoice_reference is created. They pay by regular EFT and it works.
2. EFT Client is overdue, pay by credit card, creates consolidated invoices which goes in and reverses the "regular" invoice
3. EFT Client is overdue they start the pay by credit card which creates a consolidated invoice, they eventually pay by EFT instead.
The job is smart enough to realize they had a consolidated invoice, reverses that and completes payment on the "regular" invoice

Small changes add columns etc.
EFT Friday change the jobs to match prod
Seed some data ahead of time
Cleanup test statements

File: pay-api/tests/unit/services/test_statement.py [hunk 10]
```diff
@@ -227,8 +227,9 @@ def test_get_interim_statement_change_away_from_eft(session, admin_users_mock):
     invoice_create_date = localize_date(datetime(2023, 10, 9, 12, 0))
     monthly_invoice = factory_invoice(payment_account=account, created_on=invoice_create_date,
```
Comment (seeker25) [inline pay-api/tests/unit/services/test_statement.py:228]: I don't think this was correct before?

File: jobs/payment-jobs/tasks/eft_task.py [hunk 11]
```diff
@@ -248,8 +249,29 @@ def _create_receipt_and_invoice(cls,
...
+            invoice_response = CFSService.get_invoice(cfs_account=cfs_account,
+                                                      inv_number=original_invoice_reference.invoice_number)
+            cfs_total = Decimal(invoice_response.get('total', '0'))
+            invoice_total_matches = cfs_total == invoice.total
+            if not invoice_total_matches:
+                raise ValueError(f'SBC-PAY Invoice total {invoice.total} does not match CFS total {cfs_total}')
+            # Note we do the opposite of this in payment_account.
+            current_app.logger.info(f'Consolidated invoice found, reversing consolidated '
+                                    f'invoice {invoice_reference.invoice_number}.')
+            CFSService.reverse_invoice(invoice_reference.invoice_number)
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/eft_task.py:270]: Should probably use the same code from unlock, I think this is singular, the other route handles multiples

**Review by ochiu — APPROVED** [review]
Looks good to go to me. Thanks

**Pull request merged**